### PR TITLE
Add expanded THD unit tests

### PR DESCRIPTION
**Pull request by pstjohn**
Will need to carefully xfail these on L4 test runners.

## Summary by CodeRabbit

- **Tests**
  - Expanded test coverage to validate behavior across multiple attention backends.
  - Added checks comparing losses, logits, and gradients between implementations for consistency.
  - Replaced hard-coded samples with parameterized inputs to broaden scenario coverage.
  - Added safeguards for known hardware-specific limitations to prevent false failures.
  - Improved reliability and maintainability of model validation through unified input preparation and backend toggling.

### CODE REVIEW

File: bionemo-recipes/models/esm2/tests/test_thd.py [hunk 1]
```diff
@@ -90,19 +96,101 @@ def test_thd_values_match(te_model_checkpoint, tokenizer):
...
+    bshd_grads = {name: p.grad for name, p in model_bshd.named_parameters() if p.grad is not None}
+
+    # For some reason, the word embeddings grads have a slightly higher numerical error.
+    thd_word_embeddings_grad = thd_grads.pop("esm.embeddings.word_embeddings.weight")
+    bshd_word_embeddings_grad = bshd_grads.pop("esm.embeddings.word_embeddings.weight")
+
+    torch.testing.assert_close(thd_grads, bshd_grads)
+
+    # sus
+    torch.testing.assert_close(thd_word_embeddings_grad, bshd_word_embeddings_grad, atol=1e-3, rtol=1e-5)
```
Comment (pstjohn) [inline bionemo-recipes/models/esm2/tests/test_thd.py]: well, these tests all pass on H100s. Probably need to dig into this line a bit more

Comment (trvachov) [reply bionemo-recipes/models/esm2/tests/test_thd.py]: Is it possible we co-train THD vs non THD and literally run this test every iteration?

Comment (pstjohn) [reply bionemo-recipes/models/esm2/tests/test_thd.py]: yeah we could try something like that; run it for a few iterations
